File: 103-BinaryTreeZigzagLevelOrderTraversal/103-BinaryTreeZigzagLevelOrderTraversal.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        queue, zigZagArr, rightDirection = deque([root]), [], True

        while queue:
            curr_level_numbers = []
            for i in range(len(queue)):
                currNode = queue.popleft()
                curr_level_numbers.append(currNode.val)
                if currNode.right: queue.append(currNode.right)
                if currNode.left: queue.append(currNode.left)
            if rightDirection:
                curr_level_numbers.reverse()
        
            zigZagArr.append(curr_level_numbers)
            rightDirection = not rightDirection

        return zigZagArr    
        

File: 103-BinaryTreeZigzagLevelOrderTraversal/103-BinaryTreeZigzagLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if(root == null) return [];

    let queue = [root], ans = [];
    let rightToLeft = false;

    while(queue.length){
        let nextQueue = [], res = [];
        for(let i = 0; i < queue.length; i++){
            let currNode = queue[i];
            res.push(currNode.val);

            if(currNode.left) nextQueue.push(currNode.left);
            if(currNode.right) nextQueue.push(currNode.right);
        }

        if(rightToLeft){
            res.reverse();
        }

        ans.push(res);
        rightToLeft = !rightToLeft
        queue = nextQueue;
    }
    
    return ans;
};